File: frontend/website/assets/js/diet-precedence.js
/*!
 * ZITLAS — THE diet precedence rule (assets/js/diet-precedence.js)
 *
 * Which diet is the athlete's current diet? One answer, shared with the app:
 * mobile/lib/features/diet/diet_precedence.dart implements the identical
 * rule, and both are pinned by ONE fixture —
 * tests/fixtures/diet_precedence_cases.json — run by the JS and Dart tests.
 *
 *   1. 'coach'     — the active Personal Coaching diet, when ALL hold:
 *                      relationship status 'active';
 *                      its end date (endDateTs, else endDate) absent or in
 *                        the future;
 *                      it covers diet (planType diet / complete; a free
 *                        trial's null planType is full coverage);
 *                      coaching_plans/{uid}.diet has at least one meal;
 *                      the plan belongs to the current coach (a stored
 *                        coachId that differs from the relationship's is
 *                        a previous coach's leftover);
 *                      NOT (plan planId AND live planId both present AND
 *                        different) — a null planId is allowed.
 *   2. 'expert'    — users/{uid}.dietPlan is an accepted/applied expert-
 *                    reviewed wrapper, stamped with a planId that does not
 *                    contradict the live one (an UNSTAMPED expert layer
 *                    cannot prove which goal it belongs to and is refused).
 *   3. 'ai'        — users/{uid}.dietPlan as the AI plan (wrapper without an
 *                    expert layer, or a legacy flat {days:[...]}).
 *   4. 'ai_master' — users/{uid}.dietPlanMaster, the recovery copy.
 *   5. 'none'.
 *
 * SERVER DATA ONLY. Callers pass Firestore state (live listeners / the
 * hydrated users/{uid} mirror); a localStorage copy is never an input, so a
 * stale cache can never outrank what the server holds. Timestamps are not
 * compared: the relationship decides, which is deterministic across clients
 * and clocks.
 */
(function (root) {
  'use strict';

  var DIET_PLAN_TYPES = { diet: true, complete: true };

  function toDate(v) {
    if (!v) return null;
    if (v instanceof Date) return isNaN(v.getTime()) ? null : v;
    if (typeof v.toDate === 'function') {
      try { var d0 = v.toDate(); return isNaN(d0.getTime()) ? null : d0; } catch (_) { return null; }
    }
    if (typeof v.seconds === 'number') return new Date(v.seconds * 1000);
    if (typeof v === 'string') {
      var d = new Date(v);
      return isNaN(d.getTime()) ? null : d;
    }
    return null;
  }

  function text(v) {
    return (typeof v === 'string' && v.trim()) ? v.trim() : null;
  }

  function mealsOf(day) {
    var m = day && day.meals;
    if (Array.isArray(m)) return m;
    if (m && typeof m === 'object') return Object.keys(m).map(function (k) { return m[k]; });
    return [];
  }

  /* A meal counts when it has a name — the same test both clients' parsers
     apply before rendering a coach meal at all. */
  function countMeals(plan) {
    var days = plan && Array.isArray(plan.days) ? plan.days : [];
    var n = 0;
    days.forEach(function (d) {
      mealsOf(d).forEach(function (meal) {
        if (meal && text(meal.name || meal.meal_name)) n++;
      });
    });
    return n;
  }

  function hasDays(plan) {
    return !!(plan && Array.isArray(plan.days) && plan.days.length);
  }

  /* The plan-id safety rule: only two PRESENT ids that differ disagree. */
  function contradicts(a, b) {
    a = text(a); b = text(b);
    return !!(a && b && a !== b);
  }

  function isWrapper(p) {
    return !!(p && typeof p === 'object' && p.originalDietPlan && p.currentDietPlan);
  }

  function hasExpertLayer(w) {
    if (w.isExpertPlan === true) return true;
    var mods = w.expertModifications;
    if (!mods || typeof mods !== 'object') return false;
    return Object.keys(mods).some(function (k) {
      var v = mods[k];
      return !!(v && typeof v === 'object' && Object.keys(v).length);
    });
  }

  /* Raw Firestore-shaped state -> the handful of facts the rule reads. */
  function normalize(state) {
    state = state || {};
    var rel = state.relationship || null;
    var cp = state.coachingPlan || null;
    var diet = cp && cp.diet;
    var dp = state.dietPlan || null;
    var master = state.dietPlanMaster || null;
    /* Masters are written as {planId, plan: {days}}; tolerate a flat one. */
    var masterPlan = master
      ? (master.plan && Array.isArray(master.plan.days) ? master.plan : master)
      : null;
    var wrapper = isWrapper(dp) ? dp : null;
    return {
      now: toDate(state.now) || new Date(),
      livePlanId: text(state.livePlanId),
      rel: rel ? {
        status: rel.status || null,
        coachId: text(rel.coachId),
        planType: text(rel.planType),
        end: toDate(rel.endDateTs) || toDate(rel.endDate),
      } : null,
      coach: cp ? {
        coachId: text(cp.coachId),
        planId: diet ? text(diet.planId) : null,
        meals: countMeals(diet),
      } : null,
      wrapper: wrapper ? {
        planId: text(wrapper.planId),
        expert: hasExpertLayer(wrapper),
        hasDays: hasDays(wrapper.currentDietPlan) || hasDays(wrapper.originalDietPlan),
      } : null,
      legacyFlat: !wrapper && hasDays(dp),
      master: masterPlan && hasDays(masterPlan) ? { planId: text(master.planId) } : null,
    };
  }

  function coachActiveN(n) {
    var r = n.rel, c = n.coach;
    if (!r || r.status !== 'active') return false;
    if (r.end && r.end.getTime() <= n.now.getTime()) return false;
    if (!DIET_PLAN_TYPES[r.planType || 'complete']) return false;
    if (!c || c.meals < 1) return false;
    if (c.coachId && r.coachId && c.coachId !== r.coachId) return false;
    if (contradicts(c.planId, n.livePlanId)) return false;
    return true;
  }

  /* 'expert' | 'ai' | null — the users/{uid}.dietPlan wrapper's standing. */
  function wrapperVerdictN(n) {
    var w = n.wrapper;
    if (!w || !w.hasDays) return null;
    if (contradicts(w.planId, n.livePlanId)) return null;
    if (w.expert) return w.planId ? 'expert' : null;
    return 'ai';
  }

  function selectDietSource(state) {
    var n = normalize(state);
    if (coachActiveN(n)) return 'coach';
    var verdict = wrapperVerdictN(n);
    if (verdict) return verdict;
    if (n.legacyFlat) return 'ai';
    if (n.master && !contradicts(n.master.planId, n.livePlanId)) return 'ai_master';
    return 'none';
  }

  var api = {
    selectDietSource: selectDietSource,
    coachDietActive: function (state) { return coachActiveN(normalize(state)); },
    wrapperVerdict: function (state) { return wrapperVerdictN(normalize(state)); },
    normalize: normalize,
  };
  root.ZitlasDietPrecedence = api;
})(typeof window !== 'undefined' ? window : this);

File: mobile/lib/features/diet/diet_precedence.dart
/// THE diet precedence rule — which diet is the athlete's current diet.
///
/// Line-for-line the website's `frontend/website/assets/js/diet-precedence.js`,
/// and pinned by ONE shared fixture, `tests/fixtures/diet_precedence_cases.json`,
/// run by both `test/diet_precedence_parity_test.dart` and the website's JS
/// test — so for identical Firestore state the two clients cannot disagree.
///
///  1. [DietSource.coach] — the active Personal Coaching diet, when ALL hold:
///     relationship status `active`; its end date absent or in the future;
///     it covers diet (planType `diet` / `complete`; a free trial's null
///     planType is full coverage); `coaching_plans/{uid}.diet` has at least
///     one meal; the plan belongs to the current coach; and NOT (plan planId
///     AND live planId both present AND different) — a null planId is fine.
///  2. [DietSource.expert] — `users/{uid}.dietPlan` is an accepted/applied
///     expert-reviewed wrapper stamped with a planId that does not contradict
///     the live one (an UNSTAMPED expert layer cannot prove its goal).
///  3. [DietSource.ai] — `users/{uid}.dietPlan` as the AI plan.
///  4. [DietSource.aiMaster] — `users/{uid}.dietPlanMaster`.
///  5. [DietSource.none].
///
/// Server data only — never a local cache. No timestamps are compared: the
/// relationship decides, which is deterministic across clients and clocks.
library;

import 'package:cloud_firestore/cloud_firestore.dart' show Timestamp;

enum DietSource { coach, expert, ai, aiMaster, none }

extension DietSourceWire on DietSource {
  /// The fixture's / website's spelling.
  String get wire => switch (this) {
        DietSource.coach => 'coach',
        DietSource.expert => 'expert',
        DietSource.ai => 'ai',
        DietSource.aiMaster => 'ai_master',
        DietSource.none => 'none',
      };
}

const _dietPlanTypes = {'diet', 'complete'};

String? _text(Object? v) {
  if (v is! String) return null;
  final t = v.trim();
  return t.isEmpty ? null : t;
}

DateTime? _date(Object? v) {
  if (v is DateTime) return v;
  if (v is Timestamp) return v.toDate();
  if (v is String) return DateTime.tryParse(v);
  if (v is Map && v['seconds'] is num) {
    return DateTime.fromMillisecondsSinceEpoch(((v['seconds'] as num) * 1000).round(), isUtc: true);
  }
  return null;
}

List<Object?> _mealsOf(Object? day) {
  if (day is! Map) return const [];
  final m = day['meals'];
  if (m is List) return m;
  if (m is Map) return m.values.toList();
  return const [];
}

/// A meal counts when it has a name — the same test both clients' parsers
/// apply before rendering a coach meal at all.
int _countMeals(Object? plan) {
  if (plan is! Map || plan['days'] is! List) return 0;
  var n = 0;
  for (final day in plan['days'] as List) {
    for (final meal in _mealsOf(day)) {
      if (meal is Map && _text(meal['name'] ?? meal['meal_name']) != null) n++;
    }
  }
  return n;
}

bool _hasDays(Object? plan) => plan is Map && plan['days'] is List && (plan['days'] as List).isNotEmpty;

/// The plan-id safety rule: only two PRESENT ids that differ disagree.
bool _contradicts(String? a, String? b) => a != null && b != null && a != b;

/// The standing of a `users/{uid}.dietPlan` wrapper.
class WrapperFacts {
  const WrapperFacts({this.planId, this.expert = false, this.hasDays = false});

  final String? planId;
  final bool expert;
  final bool hasDays;
}

/// The handful of facts the rule reads.
class DietPrecedenceInput {
  const DietPrecedenceInput({
    required this.now,
    this.livePlanId,
    this.hasRelationship = false,
    this.relStatus,
    this.relCoachId,
    this.relPlanType,
    this.relEnd,
    this.hasCoachPlan = false,
    this.coachPlanCoachId,
    this.coachPlanPlanId,
    this.coachMealCount = 0,
    this.wrapper,
    this.legacyFlat = false,
    this.hasMaster = false,
    this.masterPlanId,
  });

  final DateTime now;
  final String? livePlanId;
  final bool hasRelationship;
  final String? relStatus;
  final String? relCoachId;
  final String? relPlanType;
  final DateTime? relEnd;
  final bool hasCoachPlan;
  final String? coachPlanCoachId;
  final String? coachPlanPlanId;
  final int coachMealCount;
  final WrapperFacts? wrapper;
  final bool legacyFlat;
  final bool hasMaster;
  final String? masterPlanId;

  /// Raw Firestore-shaped state — exactly what the shared fixture holds and
  /// what the website's `normalize()` reads.
  factory DietPrecedenceInput.fromRaw({
    Map<String, dynamic>? relationship,
    Map<String, dynamic>? coachingPlan,
    Object? dietPlan,
    Object? dietPlanMaster,
    Object? livePlanId,
    Object? now,
  }) {
    final rel = relationship;
    final cp = coachingPlan;
    final diet = cp?['diet'];
    final dp = dietPlan is Map ? dietPlan.cast<String, dynamic>() : null;
    final master = dietPlanMaster is Map ? dietPlanMaster.cast<String, dynamic>() : null;
    final masterPlan = master == null
        ? null
        : (master['plan'] is Map && (master['plan'] as Map)['days'] is List ? master['plan'] : master);
    final isWrapper = dp != null && dp['originalDietPlan'] != null && dp['currentDietPlan'] != null;
    WrapperFacts? wrapper;
    if (isWrapper) {
      final mods = dp['expertModifications'];
      final expert = dp['isExpertPlan'] == true ||
          (mods is Map && mods.values.any((v) => v is Map && v.isNotEmpty));
      wrapper = WrapperFacts(
        planId: _text(dp['planId']),
        expert: expert,
        hasDays: _hasDays(dp['currentDietPlan']) || _hasDays(dp['originalDietPlan']),
      );
    }
    return DietPrecedenceInput(
      now: _date(now) ?? DateTime.now(),
      livePlanId: _text(livePlanId),
      hasRelationship: rel != null,
      relStatus: rel?['status'] as String?,
      relCoachId: _text(rel?['coachId']),
      relPlanType: _text(rel?['planType']),
      relEnd: _date(rel?['endDateTs']) ?? _date(rel?['endDate']),
      hasCoachPlan: cp != null,
      coachPlanCoachId: _text(cp?['coachId']),
      coachPlanPlanId: diet is Map ? _text(diet['planId']) : null,
      coachMealCount: _countMeals(diet),
      wrapper: wrapper,
      legacyFlat: !isWrapper && _hasDays(dp),
      hasMaster: _hasDays(masterPlan),
      masterPlanId: _text(master?['planId']),
    );
  }
}

bool coachDietActive(DietPrecedenceInput n) {
  if (!n.hasRelationship || n.relStatus != 'active') return false;
  final end = n.relEnd;
  if (end != null && !end.isAfter(n.now)) return false;
  if (!_dietPlanTypes.contains(n.relPlanType ?? 'complete')) return false;
  if (!n.hasCoachPlan || n.coachMealCount < 1) return false;
  if (n.coachPlanCoachId != null && n.relCoachId != null && n.coachPlanCoachId != n.relCoachId) {
    return false;
  }
  if (_contradicts(n.coachPlanPlanId, n.livePlanId)) return false;
  return true;
}

/// [DietSource.expert] / [DietSource.ai] / null — the wrapper's standing.
DietSource? wrapperVerdict(DietPrecedenceInput n) {
  final w = n.wrapper;
  if (w == null || !w.hasDays) return null;
  if (_contradicts(w.planId, n.livePlanId)) return null;
  if (w.expert) return w.planId != null ? DietSource.expert : null;
  return DietSource.ai;
}

DietSource selectDietSource(DietPrecedenceInput n) {
  if (coachDietActive(n)) return DietSource.coach;
  final verdict = wrapperVerdict(n);
  if (verdict != null) return verdict;
  if (n.legacyFlat) return DietSource.ai;
  if (n.hasMaster && !_contradicts(n.masterPlanId, n.livePlanId)) return DietSource.aiMaster;
  return DietSource.none;
}
